prepare for v0.104.0 (#5085)
signalfx/splunk-otel-collector: CHANGELOG.md

- (Core) `otlpreceiver`: Switch to `localhost` as the default for all endpoints. ([#8510](https://github.com/open-telemetry/opentelemetry-collector/pull/8510))
  Disable the `component.UseLocalHostAsDefaultHost` feature gate to temporarily get the previous default.

- (Splunk) `discovery`: When setting properties for discovery receiver as environment variables (`SPLUNK_DISCOVERY_*`), the values cannot reference other escaped environment variables without braces. For example, when trying to set discovery property `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED` to the value of another env var, `K8S_ENVIRONMENT`. For versions older than 0.104.0, setting it as `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\$K8S_ENVIRONMENT` (note the escaped variable name does not have braces) was valid. But from v0.104.0, env var names need to be passed with braces. For this example, user should modify it to `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\${K8S_ENVIRONMENT}`

> <<<<<<< HEAD
> =======
> 
> >>>>>>> d116609e7d13ccf9a9efe6671990f392ec433f30

Reviewer: ```suggestion
```

Author: bad merge from ide, cleaned

## Unreleased

## v0.104.0

This Splunk OpenTelemetry Collector release includes changes from the [opentelemetry-collector v0.104.0](https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.104.0) and the [opentelemetry-collector-contrib v0.104.0](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases/tag/v0.104.0) releases where appropriate.

> > 🚩 When setting properties for discovery receiver as environment variables (`SPLUNK_DISCOVERY_*`), the values cannot reference other environment variables without curly-braces. For example, user is trying to set discovery property `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED` to the value of another env var, `K8S_ENVIRONMENT`.

Reviewer 2: Minor suggestion: I would switch the sequence of the warning and example of the discovery receiver here.

Author: done

> For versions older than 0.104.0, setting it as `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\$K8S_ENVIRONMENT` (note the escaped variable name does not have curly braces) was valid. But from v0.104.0, env var names need to be passed with braces. For this example, user should modify it to `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\${K8S_ENVIRONMENT}`.

> > :warning: In our efforts to align with [upstream's goals](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/rfcs/env-vars.md) for environment variable resolution in collector's configuration, the Splunk OpenTelemetry Collector will be dropping support for expansion of BASH-style environment variables, such as `$FOO` in the configuration in an upcoming version. Users are adviced to update collector configuration to use `${env:FOO}` instead.

Reviewer 2: ```suggestion
> :warning: In our efforts to align with the [goals](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/rfcs/env-vars.md) defined upstream for environment variable resolution in the Collector's configuration, the Splunk OpenTelemetry Collector will be dropping support for expansion of BASH-style environment variables, such as `$FOO` in the configuration in an upcoming version. Users are advised to update their Collector's configuration to use `${env:FOO}` instead.
```

- (Core) `otlpreceiver`: Switch to `localhost` as the default for all endpoints. ([#8510](https://github.com/open-telemetry/opentelemetry-collector/pull/8510))
  Disable the `component.UseLocalHostAsDefaultHost` feature gate to temporarily get the previous default.

- (Splunk) `discovery`: When setting properties for discovery receiver as environment variables (`SPLUNK_DISCOVERY_*`), the values cannot reference other escaped environment variables without braces. For example, when trying to set discovery property `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED` to the value of another env var, `K8S_ENVIRONMENT`. For versions older than 0.104.0, setting it as `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\$K8S_ENVIRONMENT` (note the escaped variable name does not have braces) was valid. But from v0.104.0, env var names need to be passed with braces. For this example, user should modify it to `SPLUNK_DISCOVERY_EXTENSIONS_k8s_observer_ENABLED=\${K8S_ENVIRONMENT}`

> - (Contrib) `vcenterreceiver`: Drops support for vCenter 6.7 ([#33607](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33607))

Reviewer 3: ```suggestion

- (Contrib) `vcenterreceiver`: Drops support for vCenter 6.7 ([#33607](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33607))
```
Nit

  This change affects the following components:
    - extension/health_check
    - receiver/jaeger
    - receiver/sapm
    - receiver/signalfx
    - receiver/splunk_hec
    - receiver/zipkin

- (Contrib) `receiver/mongodb`: Graduate receiver.mongodb.removeDatabaseAttr feature gate to stable ([#24972](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24972))

### 💡 Enhancements 💡

- (Splunk) Set Go garbage collection target percentage to 400% ([#5034](https://github.com/signalfx/splunk-otel-collector/pull/5034))
  After removal of memory_ballast extension in v0.97.0, the Go garbage collection is running more aggressively, which
  increased CPU usage and leads to reduced throughput of the collector. This change reduces the frequency of garbage
  collection cycles to improves performance of the collector for typical workloads. As a result, the collector will
  report higher memory usage, but it will be bound to the same configured limits. If you want to revert to the previous
  behavior, set the `GOGC` environment variable to `100`.

> - (Splunk) Upgrade to golang 1.21.12

Reviewer 3: ```suggestion
- (Splunk) Upgrade to golang 1.21.12 ([#5074](https://github.com/signalfx/splunk-otel-collector/pull/5074))
```
